File: core/agent_timeout.py
from __future__ import annotations

import asyncio
from typing import Any

from flowforge.core.base_agent import BaseAgent, AgentInput, AgentOutput


class AgentTimeoutError(Exception):
    def __init__(self, agent_name: str, timeout: float) -> None:
        super().__init__(f"Agent '{agent_name}' timed out after {timeout}s")
        self.agent_name = agent_name
        self.timeout = timeout


class TimeoutAgentWrapper(BaseAgent):
    def __init__(self, agent: BaseAgent, timeout: float = 300.0) -> None:
        self._agent = agent
        self._timeout = timeout
        self.name = agent.name
        self.description = agent.description
        self.default_mode = agent.default_mode

    async def execute(self, input: AgentInput) -> AgentOutput:
        try:
            return await asyncio.wait_for(self._agent.execute(input), timeout=self._timeout)
        except asyncio.TimeoutError:
            raise AgentTimeoutError(self.name, self._timeout)

    async def execute_with_context(self, input: AgentInput, context: Any) -> AgentOutput:
        try:
            return await asyncio.wait_for(
                self._agent.execute_with_context(input, context), timeout=self._timeout
            )
        except asyncio.TimeoutError:
            raise AgentTimeoutError(self.name, self._timeout)

File: packages/plugins/modes/src/agent-timeout.ts
/**
 * Per-agent execution timeout.
 *
 * Maps flowforge Python legacy core/agent_timeout.py (F25):
 * AgentTimeoutError with message `Agent 'X' timed out after Ns` and a
 * wrapper enforcing the timeout around an execute() call.
 */

export class AgentTimeoutError extends Error {
  readonly agentName: string
  readonly timeout: number

  constructor(agentName: string, timeout: number) {
    super(`Agent '${agentName}' timed out after ${timeout}s`)
    this.name = 'AgentTimeoutError'
    this.agentName = agentName
    this.timeout = timeout
  }
}

/**
 * Race `task` against a timeout. On timeout the underlying promise is
 * left running (fire-and-forget) and AgentTimeoutError is thrown —
 * mirroring asyncio.wait_for semantics for the wrapper use case.
 */
export async function withTimeout<T>(
  task: Promise<T>,
  timeoutSeconds: number,
  agentName: string,
): Promise<T> {
  let timer: ReturnType<typeof setTimeout> | undefined
  const timeout = new Promise<never>((_, reject) => {
    timer = setTimeout(() => reject(new AgentTimeoutError(agentName, timeoutSeconds)), timeoutSeconds * 1000)
  })
  try {
    return await Promise.race([task, timeout])
  } finally {
    if (timer !== undefined) clearTimeout(timer)
  }
}
